### ISSUE TITLE
Cannot establish connection from OS X to db2 on system i

### ISSUE THREAD

**Issue by jonstorer** [ibmdb/ruby-ibmdb#53]
@arvindgu Thanks for helping me out with #51! James and I worked on that for a while.

I'm having issues connecting to the database now. I expect I'll hit the license issue eventually, but for now, I can't connect. I've google this a bit, but what I've sounds seems pretty cryptic. Could you help me out?

Thanks!

```
RuntimeError: Failed to connect to [database] due to: uncaught throw :"Connection failed: [IBM][CLI Driver] SQL30081N  A communication error has been detected. Communication protocol being used: \"TCP/IP\".  Communication API being used: \"SOCKETS\".  Location where the error was detected: \"[ip-address-removed]\".  Communication function detecting the error: \"connect\".  Protocol specific error code(s): \"61\", \"*\", \"*\".  SQLSTATE=08001 SQLCODE=-30081"
```

**Comment by arvindgu** [comment]
After ibm_db gem installation, you need to specify connection details in db2dsdriver.cfg file. Default location of db2dsdriver.cfg file would be GEM_HOME/gems_rb223_rails420/gems/ibm_db-3.0.1/lib/clidriver/cfg

For more information, please refer   http://www-01.ibm.com/support/knowledgecenter/SSEPGG_10.5.0/com.ibm.swg.im.dbclient.config.doc/doc/c0054555.html?lang=en-us&cp=SSEPGG_10.5.0

**Comment by jonstorer** [comment]
@arvindgu Can we discuss that in more detail? Is it not possible to pass the arguments in through activerecord's configuration? That's the standard strategy for ActiveRecord.

Manually editing a config file deep in a gem is not very repeatable. I'm planning on deploying this software in a virtual environment where I don't have much access to the file system. Please advise. 

Also, we'd need this for the license as well. A way to pass a path to the license.

**Comment by arvindgu** [comment]
@jonstorer You can keep the db2dsdriver.cfg file and license file under your project directory. While deploying your application you can run command to
1. Copy the license file from your project directory to GEM_HOME/gems_rb223_rails420/gems/ibm_db-3.0.1/lib/clidriver/license directory.
2. Set the *DB2DSDRIVER_CFG_PATH* environment variable to the absolute path of the db2dsdriver.cfg file present in project directory.

**Comment by jonstorer** [comment]
@arvindgu I'm hoping you can help me get this solved.

First off, in rails, in order to map active record models to a table in the database, it needs to have the configuration in within the activerecord configuration file. this would be `config/database.yml`. Here's in my sample `database.yml`:

```yml
# sample
development:
  adapter: ibm_db
  authentication: server
  database: myDatabaseName
  encoding: unicode
  host: database.domain.com
  password: sekret
  port: 1234
  schema: appSchema
  username: dbUser
```

could you help with an example of what this file would look like in the db2dsdriver.cfg?

**Comment by jonstorer** [comment]
also, I don't have a `GEM_HOME` defined. I'm using rbenv to control my ruby installs, so I'd have to solve that separately. 

Can I ask why this driver needs a separate configuration file? Why can't the driver take arguments from ruby? This driver does not follow the standard conventions for adapters for active_record. Do you have plans to create a solution?

Jon

**Comment by arvindgu** [comment]
@jonstorer  Your Rails application should be able to connect to DB2 database with Rails  `config/database.yml` *without*  db2dsdriver.cfg.  Please ensure you have specified correct connection details. 
You can quickly verify via creating a sample model and saving data via rails console, please refer below link.
http://www-01.ibm.com/support/knowledgecenter/SSEPGG_10.5.0/com.ibm.swg.im.dbclient.ruby.doc/doc/t0052768.html?lang=en

The db2dsdriver.cfg is needed only when you test connection from Interactive Ruby Shell (irb). 

Before trying with db2dsdriver.cfg, I would like to know the error you get with config/database.yml.
In case of error, please capture db2trace,  format them and send to me so that I can analyze.

**Comment by jonstorer** [comment]
@arvindgu I'm still at the same error.

```ruby
bundle exec rake my_app:db2:seed_schema                                                                                                            rake aborted!
Failed to connect to [dev] due to: uncaught throw :"Connection failed: [IBM][CLI Driver] SQL30081N  A communication error has been detected. Communication protocol being used: \"TCP/IP\".  Communication API being used: \"SOCKETS\".  Location where the error was detected: \"127.0.0.1\".  Communication function detecting the error: \"connect\".  Protocol specific error code(s): \"61\", \"*\", \"*\".  SQLSTATE=08001 SQLCODE=-30081"
[path]/my_app/lib/tasks/dev.rake:6:in `block (3 levels) in <top (required)>'
UncaughtThrowError: uncaught throw :"Connection failed: [IBM][CLI Driver] SQL30081N  A communication error has been detected. Communication protocol being used: \"TCP/IP\".  Communication API being used: \"SOCKETS\".  Location where the error was detected: \"127.0.0.1\".  Communication function detecting the error: \"connect\".  Protocol specific error code(s): \"61\", \"*\", \"*\".  SQLSTATE=08001 SQLCODE=-30081"
[path]/my_app/lib/tasks/dev.rake:6:in `block (3 levels) in <top (required)>'
Tasks: TOP => my_app:db2:seed_schema
(See full trace by running task with --trace)
```

The code I'm trying to run is:

```ruby
config = YAML.load_file(Rails.root.join('config', 'database.yml'))['ibm']
connection = ActiveRecord::Base.establish_connection(config)
puts connection.connection.execute("list tables")
```

```yml
ibm: &ibm
  adapter: ibm_db
  encoding: unicode
  database: myappdev
  authentication: server
  username: db2inst1
  schema: db2inst1
  password: super-sekret
  host: localhost
  port: 50000
```

I'm attempting to connect rails to an existing db2 instance. For development I need to recreate the schema. I'm attempting to get the connection established before I do any of the harder work. I've moved on to use the ibm db2 docker image (seems to be working great, I can connect to the db2 instance in the docker image, just not externally.) However, this is the same issues I've run into attempting to connect with a local (non-docker) install.

**Comment by jonstorer** [comment]
@arvindgu Also, thanks for your quick response. I think I might be doing something wrong gathering the db2trc logs. I was able to create a 32MB file of what looked like binary. Here's a sample:

```
@TRACE@^@7.0^@^@^@^@^@Linux/X8664^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@s141128^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^B^@
^@^@^@^B^@^@^@^A^A^A^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@
^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@
^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@
^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@
^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@
```

If I'm not doing this right, please let me know how to get it done.

Thanks!

**Comment by arvindgu** [comment]
@jonstorer Below are the steps to collect db2traces

1. Open a console and run below commands and change present working directory to GEM_HOME/gems/ibm_db-3.0.1/lib/clidriver/adm

2. Run below command to start tracing
   ./db2trc on -f   /home/<user>/db2trc.dump

3. Now run application and reproducing the problem

4.Run below command to stop tracing
    ./db2trc off 
	
5. Now format the traces with below commands
	./db2trc fmt /home/<user>/db2trc.dump /home/<user>/db2trc.fmt
	./db2trc flw /home/<user>/db2trc.dump /home/<user>/db2trc.flw
	./db2trc fmt -c /home/<user>/db2trc.dump /home/<user>/db2trc.drda
	./db2trc fmt -cli /home/<user>/db2trc.dump /home/<user>/db2trc.cli 

6. Please send me .fmt .flw .drda .cli to my email id.

In case you get any problem like "dyld: Library not loaded: libdb2.dylib", please add GEM_HOME/gems/ibm_db-3.0.1/lib/clidriver/lib  path to DYLD_LIBRARY_PATH.

**Comment by arvindgu** [comment]
@jonstorer Also confirm that you are connecting to correct port, you can check the port number of DB2 server by running below command 
db2 get dbm cfg | grep SVC

**Comment by jonstorer** [comment]
I'm not getting a port back when I run this command.

```shell
[db2inst1@e766a159c593 ~]$ db2 get dbm cfg | grep SVC
 TCP/IP Service name                          (SVCENAME) = db2c_db2inst1
 SSL service name                         (SSL_SVCENAME) =
```

I'm working on getting the trace together.

**Comment by arvindgu** [comment]
As you are connecting with port number as 50000. Please run below on db2 command prompt and then try connecting with Ruby application.

db2 update dbm cfg using SVCENAME 50000
db2set db2comm=tcpip
db2stop
db2start

**Comment by jonstorer** [comment]
@arvindgu I'm still receiving the same error message with the same error codes. :-/

**Comment by arvindgu** [comment]
Can you see if you are able to connect to remote DB2 database. Also verify whether you are able to connect to local database using Ruby irb console.

**Comment by jonstorer** [comment]
@arvindgu I just attempted to connect. The connection string was copied & altered from the connection string built when connecting in a rails app.

script
```ruby
#! /usr/bin/env ruby

require('ibm_db')
IBM_DB.connect("DRIVER={IBM DB2 ODBC DRIVER}; DATABASE=mydb; HOSTNAME=my.host.example.com; PORT=1234; PROTOCOL=TCPIP; UID=myuid; PWD=mypassword; SECURITY=SSL; AUTHENTICATION=server;", "", "")

```

output
```ruby
test.rb:4:in `connect': uncaught throw :"Connection failed: [IBM][CLI Driver] SQL30081N  A communication error has been detected. Communication protocol being used: \"SSL\".  Communication API being used: \"SOCKETS\".  Location where the error was detected: \"\".  Communication function detecting the error: \"sqlccSSLSocketSetup\".  Protocol specific error code(s): \"420\", \"*\", \"*\".  SQLSTATE=08001 SQLCODE=-30081" (UncaughtThrowError)
```

**Comment by arvindgu** [comment]
I see you have specified SECURITY=SSL, check whether your DB2 server is configured with SSL. If DB2 Server is not configured with SSL, then try removing the SECURITY=SSL from connection string.

**Comment by soumithx** [comment]
still not working

**Comment by Fishy49** [comment]
I know this is old, so I'm sorry to bump, but for future-seekers. The initial error reported in this issue (with protocol-specific error 61) relates to a connection problem. For me, I was using the wrong port. I had to communicate with the admins of the DB2 instance to determine which port was being used and once that was correct - I was able to connect without issue (only to find out we needed a license but I digress).
